### PR TITLE
[IA-4910] add get endpoint

### PR DESCRIPTION
**Pull request by jdcanas**

### CODE REVIEW

File: service/src/main/java/bio/terra/appmanager/service/ChartService.java [hunk 1]
```diff
@@ -25,4 +26,9 @@ public ChartService(ChartVersionDao chartVersionDao) {
   public void createVersions(@NotNull List<ChartVersion> versions) {
     versions.forEach(chartVersionDao::upsert);
   }
+
+  @ReadTransaction
+  public List<ChartVersion> getVersions(@NotNull List<String> names, @NotNull Boolean includeAll) {
+    return chartVersionDao.get(names, includeAll);
```
Comment (cpate4) [inline service/src/main/java/bio/terra/appmanager/service/ChartService.java:32]: my two cents ...
this line is currently not covered by unit tests. i would specify in the test to mock out the dao for yourself to show it is a pass-through (or create a separate service-level test)

File: service/src/main/resources/api/openapi.yml [hunk 2]
```diff
@@ -156,6 +184,10 @@ components:
       type: array
       items:
         $ref: '#/components/schemas/ChartVersion'
+    ChartName:
+      type: string
+      pattern: '^[a-z]+(-[a-z]+)*$'
```
Comment (cpate4) [inline service/src/main/resources/api/openapi.yml:189]: i'm a big fan of adding links to reference documentation for our future selves. what do you think about either adding a link to the external documentation that says why we are implementing this?, and/or updating our DESIGN-service.md to indicate that this is how we are handling data validation (at the openapi.yml level)?

File: service/src/main/java/bio/terra/appmanager/model/ChartVersion.java [hunk 3]
```diff
@@ -28,6 +28,15 @@ public static ChartVersion fromApi(bio.terra.appmanager.api.model.ChartVersion s
...
   }
 
+  public bio.terra.appmanager.api.model.ChartVersion toApi() {
+    return new bio.terra.appmanager.api.model.ChartVersion()
+        .chartName(this.chartName)
+        .chartVersion(this.chartVersion)
+        .appVersion(this.appVersion)
+        .activeAt(this.activeAt)
+        .inactiveAt(this.inactiveAt);
+  }
```
Comment (cpate4) [inline service/src/main/java/bio/terra/appmanager/model/ChartVersion.java:38]: for me (ideally) this would be tested at some level. given it is used in the API, can we add a unit test showing the return from the service (or mock'd dao) and then consequently exercising this code to ensure completeness?

alternatively, we could implement a unit test of this single method, but I like the former better because ultimately we shouldn't care about the internal methods so much as we want to ensure the internal model that comes out out our dao is ultimately transformed into what we are expecting at the controller level

**Review by cpate4 — COMMENTED** [review]
... it's all fairly straight forward.
main comments are around testing (for completeness) and documentation (for our future selves)

File: service/src/main/java/bio/terra/appmanager/service/ChartService.java (hunk 1 above)
Comment (jdcanas) [reply service/src/main/java/bio/terra/appmanager/service/ChartService.java:32]: Ah didn't see the service test file, 👍

File: service/src/main/resources/api/openapi.yml (hunk 2 above)
Comment (jdcanas) [reply service/src/main/resources/api/openapi.yml:189]: I'm going to roll this back after some consideration, and address these concerns in a dedicated validation ticket with proper testing

File: service/src/main/java/bio/terra/appmanager/model/ChartVersion.java (hunk 3 above)
Comment (jdcanas) [reply service/src/main/java/bio/terra/appmanager/model/ChartVersion.java:38]: I added the test to `ChartVersionDaoTest`

File: service/src/test/java/bio/terra/appmanager/dao/ChartVersionDaoTest.java [hunk 4]
```diff
@@ -212,6 +212,27 @@ void testMultiDelete() {
     }
   }
 
+  @Test
+  void testChartVersionDaoToApi() {
```
Comment (cpate4) [inline service/src/test/java/bio/terra/appmanager/dao/ChartVersionDaoTest.java:216]: my recommendation would be to move this to either a `Model` or `Controller` test as the Controller is the one actually doing the conversion. you could call the `Controller` method directly to ensure it is doing what you expect vs validating the HTTP interface

it feels weird to have it in the `Dao` test layer because the `Dao` has no knowledge of the `api.schema`

File: service/src/test/java/bio/terra/appmanager/service/ChartServiceTest.java [hunk 5]
```diff
@@ -62,4 +62,19 @@ void testCreateChartVersion_multipleElement() {
...
+    Boolean includeAll = true;
+
+    ArgumentCaptor<List<String>> argument1 = ArgumentCaptor.forClass(List.class);
+    ArgumentCaptor<Boolean> argument2 = ArgumentCaptor.forClass(Boolean.class);
+
+    InOrder inOrder = inOrder(chartVersionDao);
+    chartService.getVersions(chartNameList, includeAll);
+    inOrder.verify(chartVersionDao, calls(1)).get(argument1.capture(), argument2.capture());
+    assertEquals(chartNameList, argument1.getValue());
+    assertEquals(includeAll, argument2.getValue());
```
Comment (cpate4) [inline service/src/test/java/bio/terra/appmanager/service/ChartServiceTest.java:78]:
```suggestion
    inOrder.verify(chartVersionDao, calls(1)).get(eq(chartNameList), eq(includeAll));
```

File: service/src/main/java/bio/terra/appmanager/controller/AdminController.java [hunk 6]
```diff
@@ -27,4 +28,21 @@ public ResponseEntity<Void> createChartVersions(List<ChartVersion> body) {
     this.chartService.createVersions(versions);
     return ResponseEntity.noContent().build();
   }
+
+  // Note that this method's implementation relies on `includeAll` having a default value and being
+  // not null
+  @Override
+  public ResponseEntity<ChartArray> getChartVersions(String chartName, Boolean includeAll) {
+    List<String> versions = chartName == null ? List.of() : List.of(chartName);
```
Comment (cpate4) [inline service/src/main/java/bio/terra/appmanager/controller/AdminController.java:36]: I confirmed that the code works "as expected" given the `openapi.yml` sets the default, but should we code defensively and add the following to the code?

```suggestion
  // Note that this method's implementation relies on `includeAll` having a default value and being
  // not null
  @Override
  public ResponseEntity<ChartArray> getChartVersions(String chartName, Boolean includeAll) {
    List<String> versions = chartName == null ? List.of() : List.of(chartName);
    includeAll = (includeAll != null)?(includeAll):(Boolean.FALSE)
```

File: service/src/test/java/bio/terra/appmanager/dao/ChartVersionDaoTest.java (hunk 4 above)
Comment (jdcanas) [reply service/src/test/java/bio/terra/appmanager/dao/ChartVersionDaoTest.java:216]: Fair... I misunderstood your first comment, never worked with spring mocking and the `Autowired` vs `MockBean` annotations. I think I figured it out via `  @Autowired AdminController controller;` in the appropriate file and moved the test there.

File: service/src/test/java/bio/terra/appmanager/service/ChartServiceTest.java (hunk 5 above)
Comment (jdcanas) [reply service/src/test/java/bio/terra/appmanager/service/ChartServiceTest.java:78]: Hmm... your solution is definitely more concise, but for my own learning, not sure I understand the use case of explicit `eq` matcher vs argument captors as the latter is relatively new to me.

Is an argument captor for more complex matching when you want to only verify an aspect of said complex argument rather than a more primitive argument where a direct match is feasible?

File: service/src/main/java/bio/terra/appmanager/controller/AdminController.java (hunk 6 above)
Comment (jdcanas) [reply service/src/main/java/bio/terra/appmanager/controller/AdminController.java:36]: I almost did the defensive approach myself but put a comment instead. I don't feel too strongly, I'll go the route you recommend since it is more defensive rather than rely on a future refactoring seeing the comment and having our code break if someone removes that default.

Comment (cpate4) [reply service/src/main/java/bio/terra/appmanager/controller/AdminController.java]: ... and I am six-in-one, so happy you are willing to engage the conversation with me. Thank you!

I was trying to think of a way to "ensure" behavior ... an alternative would be to write a test where we don't specify the `includeAll` and ensure it comes through as `false` - that should suffice as well. Mindset being, every line of code we write is just one more opportunity for a bug (at some point). The test will at least confirm our assumptions and it is run regularly. Thoughts?

File: service/src/test/java/bio/terra/appmanager/service/ChartServiceTest.java (hunk 5 above)
Comment (cpate4) [reply service/src/test/java/bio/terra/appmanager/service/ChartServiceTest.java]: Yes (re: more complex matching), and ...

I was originally using the `ArgumentCaptor` because I was trying to figure out how to capture the class of the same name in the `...ControllerTest` and came upon it in a StackOverflow article. I wanted to give precedence to the ChartVersion that was a part of the `api.model.` and that is how it was prescribed. 

Then, in this case, I was using it because the test had the `ChartVersion` being modified by the `...Dao`, and I only wanted to validate that the `ChartVersion` was the same (and the `eq(...)` method would have also evaluated the date columns for equality) - so that is how I ended up here. I now, too, wonder if any of my old tests could be simplified ... 🤔

File: service/src/main/java/bio/terra/appmanager/controller/AdminController.java (hunk 6 above)
Comment (jdcanas) [reply service/src/main/java/bio/terra/appmanager/controller/AdminController.java]: Yeah, I'll add a test for `null` includeAll passed to the controller and verify the service mock gets false.

Comment (cpate4) [reply service/src/main/java/bio/terra/appmanager/controller/AdminController.java]: ... and then you can just remove the `includeAll = ...`. I feel like this is the least of the smells ... thank you for bearing with me

**Review by cpate4 — APPROVED** [review]
